cdap-api/src/main/java/co/cask/cdap/api/lineage/LineageRecorder.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.lineage;
+
+import co.cask.cdap.api.lineage.operation.Operation;
+import co.cask.cdap.api.lineage.operation.Read;
+import co.cask.cdap.api.lineage.operation.Write;
+
+import java.util.Collection;
+
+/**
+ * Interface for recording the lineage information.
+ */
+public interface LineageRecorder {
```

Comment:
Should this be called FieldLineageRecorder? Otherwise it is too similar to the dataset level lineage that CDAP already records without this API, users might get confused?

Reply:
But this is only field-level. What is an input other than a field? Input and output are very generic terms and to not convey the meaning of the API. 

Input is also generic and ambiguous: suppose an operation takes a first name and a last name and concatenates them to form a name. Then input of the operation is the (first, last), the input schema is (string, string) but also first and last are each an input of the operation. 

Furthermore, we already have an Input class that describes an input for a mapreduce or spark job, and that input is an entire dataset or logical data entity. For one, we confuse users by using the same terminology. But also code will look ugly because both these classes will be used in the same place (the initialize of a program), and users will have to use fully qualified class names in their code, which I find ugly.